soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -207,12 +216,149 @@ static int dt_init(const struct device *dev)
 	return intel_adsp_ipc_init(dev);
 }
 
+#ifdef CONFIG_PM_DEVICE
+
+void intel_adsp_ipc_set_resume_handler(const struct device *dev,
+	intel_adsp_ipc_resume_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->resume_fn = fn;
+	api->resume_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+void intel_adsp_ipc_set_suspend_handler(const struct device *dev,
+	intel_adsp_ipc_suspend_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->suspend_fn = fn;
+	api->suspend_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+/**
+ * @brief Waits for IPC device to become non-busy.
+ * Peforms @see retries number of checks for the busy status each time after @see usec_to_wait
+ * elapsed.
+ *
+ * @param usec_to_wait Time to wait in microseconds before checking again for the busy status.
+ * @param retries Number of retries Number of times to repeat a busy check.
+ * @return true Returned when device became non-busy during expected timeline.
+ * @return false Returned when device is still busy when expected timeline elapsed.
+ */
+static bool wait_ipc_device_not_busy(uint32_t usec_to_wait, uint32_t retries)
+{
+	if (pm_device_is_busy(INTEL_ADSP_IPC_HOST_DEV)) {
+		while (retries > 0) {
+			k_busy_wait(usec_to_wait);
+			if (!pm_device_is_busy(INTEL_ADSP_IPC_HOST_DEV))
+				return true;
+			retries--;
+		}
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * @brief Manages IPC driver power state change.
+ *
+ * @param dev IPC device.
+ * @param action Power state to be changed to.
+ * @return int Returns 0 on success or optionaly error code from the
+ * registered ipc_power_control_api callbacks. Returning error code from functions
+ * invoked during power stansition, does not release the pm_device_state_lock.
+ * It must be released manually by the caller during error handling process.
```

Comment:
this is strange, any particular reason for this?

Reply:
No, comment I forgot to remove from older revisions of this PR. I'll clean it up.